src/interface/python/pyadiak.cpp

```diff
@@ -0,0 +1,260 @@
+bool value(std::string name, T value, int category = adiak_general,
+           std::string subcategory = "") {
+  return adiak::value(name, value.to_adiak(), category, subcategory);
+}
+
+template <class T, typename = pybind11::detail::enable_if_t<
+                       is_templated_base_of<DataContainer, T>::value>>
+bool value_vec(std::string name, const std::vector<T>& value,
+               int category = adiak_general, std::string subcategory = "") {
+  using adiak_type = decltype(std::declval<T>().to_adiak());
+  std::vector<adiak_type> adiak_vec;
+  std::transform(value.cbegin(), value.cend(), std::back_inserter(adiak_vec),
+                 [](const T& v) { return v.to_adiak(); });
+  return adiak::value(name, adiak_vec, category, subcategory);
+}
+
+// Custom version of adiak::value for sets of subclasses of DataContainer
+template <class T, typename = pybind11::detail::enable_if_t<
+                       is_templated_base_of<DataContainer, T>::value>>
+bool value_set(std::string name, const std::set<T>& value,
+               int category = adiak_general, std::string subcategory = "") {
+  using adiak_type = decltype(std::declval<T>().to_adiak());
+  std::set<adiak_type> adiak_set;
+  std::transform(value.cbegin(), value.cend(), std::inserter(adiak_set, adiak_set.end()),
+                 [](const T& v) { return v.to_adiak(); });
+  return adiak::value(name, adiak_set, category, subcategory);
+}
+
+// Macros to help generate the lambdas that pybind11 needs to bind the inline
+// functions in adiak.hpp
+//
+// clang-format off
+#define GENERATE_API_CALL_NO_ARGS(name)                                                \
+  [](){ return adiak::name(); }
+#define GENERATE_API_CALL_ONE_ARG(name, arg_type) [](arg_type val){ return adiak::name(val); }
+// clang-format on
+
+// Function to populate the Python module
+void create_adiak_annotation_mod(py::module_ &mod) {
+  // Try to import mpi4py so that we can early abort if
+  // we can't properly create bindings
+  if (import_mpi4py() < 0) {
+    throw std::runtime_error(
+        "Cannot load mpi4py within the Adiak Python bindings");
+  }
+
+  // Bind 'init' and 'fini'
+  mod.def("init", &adiak::python::init);
+  mod.def("fini", GENERATE_API_CALL_NO_ARGS(fini));
+  // Python has no concept of templates, so pybind11 requires
+  // us to specify the types for all template functions. The requires us to
+  // iterate over all acceptable value types for adiak::value.
+  //
+  // Because we need to use Python-compatible types everywhere,
+  // we use special wrappers for types not natively supported by pybind11.
+  // To bridge these types to Adiak, we use custom versions of "value".
+  //
+  // We map all these specializations to the Python "value" function.
+  // Pybind11 will do the work of selecting the correct version of this
+  // function for us.
+  mod.def("value", [](const std::string& name, int8_t val, int category, const std::string& subcategory) {
```

Comment:
Not sure why we there is only `int8_t` of all things? Should this be `int64_t` or just regular `int` instead?

Reply:
[fb282ab](https://github.com/LLNL/Adiak/pull/38/commits/fb282ab4481c14e9e4b68eaf50dcecc215364c1e) @ilumsden Was there a specific reason this was `int8_t` that we should be aware of?